linker: add no-flash system support

wentongwu opened the pull request:
add NO_FLASH configure option and no-flash system makes text,
rodata, and data all in SRAM, so it's also a non-XIP system.
And when NO_FLASH is enabled, XIP should be disable and vice versa.

Also make code relocation work for non-XIP system.

ioannisg reviewed (commented):
Would we need this extra Kconfig, or could we live with overwriting XIP? (CONFIG_XIP=n)

wentongwu commented:
> Would we need this extra Kconfig, or could we live with overwriting XIP? (CONFIG_XIP=n)
Thanks for the review. Yes, when doing this PR, I thought no-flash system should be equal to non-XIP system, but I saw this "#if defined(CONFIG_XIP) && (FLASH_SIZE != 0)" in include/arch/arc/v2/linker.ld
, so I think no-flash system should be non-XIP system, but I'm not sure if non-XIP system is no-flash system. And I imagine a system which has FLASH, but the bootloader will help copy text, rodata, and data to SRAM, for Zephyr, it's a non-XIP system, but it has image located in FLASH and in linker script we should hold the FLASH memory block. I know the bootloader is redundant, but I don't want to break any hardware which I don't know with the change. And if non-XIP system is no-flash system, I would like to change the code.

tyrantlizard77 reviewed (commented):
this patch looks OK to me, but only defines this stuff for ARM. How hard would it be to make this work for all arches?

wentongwu commented:
> this patch looks OK to me, but only defines this stuff for ARM. How hard would it be to make this work for all arches?

I'm checking other arch's linker scripts.

wentongwu commented:
> this patch looks OK to me, but only defines this stuff for ARM. How hard would it be to make this work for all arches?

I see other arches have already considered no-flash or all SRAM satiation.  Only arc arch has distinguish with no-flash and non-XIP. Others have FLASH/ROM wrapped with CONFIG_XIP. But arm has nothing with FLASH. For now I'm not familiar with other arches, so I want to change arm arch first, is that ok? @andrewboie

ioannisg commented on arch/arm/core/cortex_m/Kconfig:
```diff
@@ -317,8 +317,16 @@ config FAULT_DUMP
+	default n
+	help
+	  The image of no FLASH system will be downloaded to SRAM by other
+	  big system(like some of Wi-Fi chipsets in Anroid cellphone).
```
How is this description different from a description of a Non-XIP system?

ioannisg reviewed (commented):
> 
> 
> > Would we need this extra Kconfig, or could we live with overwriting XIP? (CONFIG_XIP=n)
> 
> @ioannisg Thanks for the review. Yes, when doing this PR, I thought no-flash system should be equal to non-XIP system, but I saw this "#if defined(CONFIG_XIP) && (FLASH_SIZE != 0)" in include/arch/arc/v2/linker.ld
> , so I think no-flash system should be non-XIP system, but I'm not sure if non-XIP system is no-flash system. And I imagine a system which has FLASH, but the bootloader will help copy text, rodata, and data to SRAM, for Zephyr, it's a non-XIP system, but it has image located in FLASH and in linker script we should hold the FLASH memory block. I know the bootloader is redundant, but I don't want to break any hardware which I don't know with the change. And if non-XIP system is no-flash system, I would like to change the code.

A no-flash system is a non-XIP system, correct.

I am trying to understand: if we build an image (without your patch) with CONFIG_XIP=n and FLASH_SIZE=0, is it any different from building an image (with your patch) with CONFIG_NO_FLASH=y?

wentongwu commented:
> > > Would we need this extra Kconfig, or could we live with overwriting XIP? (CONFIG_XIP=n)
> > 
> > 
> > @ioannisg Thanks for the review. Yes, when doing this PR, I thought no-flash system should be equal to non-XIP system, but I saw this "#if defined(CONFIG_XIP) && (FLASH_SIZE != 0)" in include/arch/arc/v2/linker.ld
> > , so I think no-flash system should be non-XIP system, but I'm not sure if non-XIP system is no-flash system. And I imagine a system which has FLASH, but the bootloader will help copy text, rodata, and data to SRAM, for Zephyr, it's a non-XIP system, but it has image located in FLASH and in linker script we should hold the FLASH memory block. I know the bootloader is redundant, but I don't want to break any hardware which I don't know with the change. And if non-XIP system is no-flash system, I would like to change the code.
> 
> A no-flash system is a non-XIP system, correct.
> 
> I am trying to understand: if we build an image (without your patch) with CONFIG_XIP=n and FLASH_SIZE=0, is it any different from building an image (with your patch) with CONFIG_NO_FLASH=y?

it's same, I think CONFIG_XIP=n and FLASH_SIZE=0 has same meaning with NO_FLASH, both of them define no-flash system, and we won't add another Kconfig if with CONFIG_XIP=n and FLASH_SIZE=0. 

But we should also add below changes if using CONFIG_XIP=n and FLASH_SIZE=0:

+#if (CONFIG_XIP==n)&& (FLASH_SIZE == 0)
+#define ROM_ADDR RAM_ADDR
+#else
#define ROM_ADDR (CONFIG_FLASH_BASE_ADDRESS + CONFIG_FLASH_LOAD_OFFSET)
+#endif

If above one is ok, I will re-make the patch. @ioannisg

wentongwu committed:
linker: add no-flash system support

Non-XIP system with FLASH_SIZE = 0 is no-flash system. And no-flash
system makes text, rodata, and data all in SRAM, so define the marco
ROM_ADDR to RAM_ADDR.

Fixes: #16027.

Signed-off-by: Wentong Wu <wentong.wu@intel.com>

gnuless commented:
> A no-flash system is a non-XIP system, correct.

But a non-XIP system is not *necessarily* a non-flash system.  These two configurations might correlate now, but they're not inherently linked.  I wonder if perhaps @wentongwu was on the right track decoupling the concepts here before he was nudged back the other way.

ioannisg commented:
> 
> 
> > A no-flash system is a non-XIP system, correct.
> 
> But a non-XIP system is not _necessarily_ a non-flash system. These two configurations might correlate now, but they're not inherently linked. I wonder if perhaps @wentongwu was on the right track decoupling the concepts here before he was nudged back the other way.

I am not trying to achieve ideal definitions here. All I push for is to use the minimum number of Kconfig options to describe our use-cases, if we can, in an effort to reduce complexity.

From Zephyr's perspective, I am asking ourselves:
-  why would a non-XIP  build define a FLASH_SIZE other than zero?
- can we achieve the desired result with setting FLASH and XIP options  to zero and not-defined, respectively?

If we need, in the future to cover more use-cases, we can consider extending the Kconfig state space.

gnuless commented:
> I am not trying to achieve ideal definitions here. All I push for is to use the minimum number of Kconfig options to describe our use-cases, if we can, in an effort to reduce complexity.

While I agree that Kconfig options should be kept to a "minimum" (whatever that actually means), but unnecessarily coupling options to reduce their number is simply not good software engineering.  The mere fact that there's been a back-and-forth discussion over what the options in question here even mean, and how they interact, is a good indication of that.  But I agree that this patch is not the time or place to attempt to correct this.
